### PR TITLE
[Feature] Firestorm supports MapReduce (Part 3 - Reduce Side)

### PR DESCRIPTION
**Pull request by frankliee**
### What changes were proposed in this pull request?
We add a new ShuffleConsumerPlugin to support the using rss in MapReduce. 

### Why are the changes needed?
In real production environments, there are too many MapReduce's jobs. The work is meaningful.

### Does this PR introduce _any_ user-facing change?
Yes, we bring some new configurations. In another pr, we will update the doc about the use of this feature.

### How was this patch tested?
new Uts.

### CODE REVIEW

File: client-mr/src/main/java/org/apache/hadoop/mapreduce/v2/app/RssMRAppMaster.java [hunk 1]
```diff
@@ -134,6 +134,13 @@ public static void main(String[] args) {
       }
       conf.set(RssMRConfig.RSS_ASSIGNMENT_PREFIX + entry.getKey(), StringUtils.join(servers, ","));
     });
+
+    // close slow start
+    if(conf.getFloat(MRJobConfig.COMPLETED_MAPS_FOR_REDUCE_SLOWSTART, 0.05f) != 1) {
+      conf.set(MRJobConfig.COMPLETED_MAPS_FOR_REDUCE_SLOWSTART, "1");
+      LOG.info("close slow start, because RSS does not support it yet");
```
Comment (jerqi) [inline client-mr/src/main/java/org/apache/hadoop/mapreduce/v2/app/RssMRAppMaster.java:141]: WARN

File: client-mr/pom.xml [hunk 2]
```diff
@@ -90,6 +95,10 @@
             <artifactId>protobuf-java</artifactId>
             <version>${protobuf.version}</version>
         </dependency>
+        <dependency>
+            <groupId>com.tencent.rss</groupId>
+            <artifactId>rss-client-spark-common</artifactId>
```
Comment (jerqi) [inline client-mr/pom.xml:100]: remove

File: client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssShuffle.java [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+import org.apache.hadoop.mapred.*;
+import org.apache.hadoop.mapreduce.MRRssUtils;
+import org.apache.hadoop.mapreduce.RssMRConfig;
+import org.apache.hadoop.mapreduce.TaskAttemptID;
+import org.apache.hadoop.util.Progress;
+import org.apache.hadoop.yarn.api.ApplicationConstants;
+import org.apache.hadoop.yarn.api.records.ContainerId;
+import org.roaringbitmap.longlong.Roaring64NavigableMap;
+
+public class RssShuffle<K, V> implements ShuffleConsumerPlugin<K, V>, ExceptionReporter {
```
Comment (jerqi) [inline client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssShuffle.java:43]: RssShuffleConsumer？

File: client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssBypassWriter.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+import org.apache.commons.logging.LogFactory;
+import org.apache.hadoop.io.IOUtils;
+import org.apache.hadoop.util.ReflectionUtils;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.lang.reflect.Field;
+import java.nio.ByteBuffer;
+
+public class RssBypassWriter {
```
Comment (jerqi) [inline client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssBypassWriter.java:32]: Why do we need write to disk?

File: client-mr/src/main/java/org/apache/hadoop/mapreduce/v2/app/RssMRAppMaster.java (hunk 1 above)
Comment (frankliee) [reply client-mr/src/main/java/org/apache/hadoop/mapreduce/v2/app/RssMRAppMaster.java:141]: ok

File: client-mr/pom.xml (hunk 2 above)
Comment (frankliee) [reply client-mr/pom.xml:100]: ok

File: client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssShuffle.java (hunk 3 above)
Comment (frankliee) [reply client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssShuffle.java:43]: In original MR， this class is just called `Shuffle` (does not ask me why).

File: client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssBypassWriter.java (hunk 4 above)
Comment (frankliee) [reply client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssBypassWriter.java:32]: MR's merger will allocate to write very larger mapout to disk directly.

File: client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssEventFetcher.java [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+      fromEventIdx += events.length;
+
+      // Process the TaskCompletionEvents:
+      // 1. Save the SUCCEEDED maps in knownOutputs to fetch the outputs.
+      // 2. Save the OBSOLETE/FAILED/KILLED maps in obsoleteOutputs to stop
+      //    fetching from those maps.
+      // 3. Remove TIPFAILED maps from neededOutputs since we don't need their
+      //    outputs at all.
+      for (TaskCompletionEvent event : events) {
+        scheduler.resolve(event);
```
Comment (jerqi) [inline client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssEventFetcher.java:144]: We don't think we should add the concept `scheduler`.

File: client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssShuffle.java [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+    // start shuffle fetcher
+    CreateShuffleReadClientRequest request = new CreateShuffleReadClientRequest(
+      appId, 0, reduceId.getTaskID().getId(), storageType, basePath, indexReadLimit, readBufferSize,
+      partitionNumPerRange, partitionNum, blockIdBitmap, taskIdBitmap, serverInfoList, jobConf);
+    ShuffleReadClient shuffleReadClient = ShuffleClientFactory.getInstance().createShuffleReadClient(request);
+    RssFetcher fetcher = new RssFetcher(jobConf, reduceId, scheduler, merger, reporter, metrics, this, shuffleReadClient);
+    fetcher.start();
+
+    // Wait for shuffle to complete successfully
+    while (!scheduler.waitUntilDone(PROGRESS_FREQUENCY)) {
```
Comment (jerqi) [inline client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssShuffle.java:175]: Don't use the process like MR, it's bad design. It's hard to understand. It's not necessary to use multi-thread.

Comment (frankliee) [reply client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssShuffle.java:175]: Ok

File: client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssEventFetcher.java (hunk 5 above)
Comment (frankliee) [reply client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssEventFetcher.java:144]: scheduler determines which task are required, and upload the progress.

File: client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssShuffle.java (hunk 3 above)
Comment (jerqi) [reply client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssShuffle.java:43]: OK, although a little strange.

File: client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssBypassWriter.java (hunk 4 above)
Comment (jerqi) [reply client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssBypassWriter.java:32]: Could we use merger's code?

Comment (frankliee) [reply client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssBypassWriter.java:32]: merger's code is difficult to use.

File: client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssEventFetcher.java (hunk 5 above)
Comment (jerqi) [reply client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssEventFetcher.java:144]: It's not neccessary concept. We can write a method to implement the function.

File: client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssBypassWriter.java (hunk 4 above)
Comment (jerqi) [reply client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssBypassWriter.java:32]: Add some comments.

Comment (frankliee) [reply client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssBypassWriter.java:32]: Done

File: client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssEventFetcher.java (hunk 5 above)
Comment (frankliee) [reply client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssEventFetcher.java]: scheduler has been removed.

File: client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssShuffle.java (hunk 6 above)
Comment (frankliee) [reply client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssShuffle.java]: The multi-threaded fashion is updated single-threaded fashion.

**Comment by frankliee** [comment]
The map side has been rebased.

File: common/pom.xml [hunk 7]
```diff
@@ -68,5 +68,10 @@
       <groupId>org.roaringbitmap</groupId>
       <artifactId>RoaringBitmap</artifactId>
     </dependency>
+    <dependency>
+      <groupId>net.jpountz.lz4</groupId>
+      <artifactId>lz4</artifactId>
+      <version>1.3.0</version>
```
Comment (jerqi) [inline common/pom.xml:74]: we manage the version in the parent pom file.

File: client-mr/src/main/java/org/apache/hadoop/mapreduce/MRRssUtils.java [hunk 8]
```diff
@@ -0,0 +1,86 @@
+/*
+ * Tencent is pleased to support the open source community by making
```
Comment (jerqi) [inline client-mr/src/main/java/org/apache/hadoop/mapreduce/MRRssUtils.java:2]: I renamed the file name.

File: client-mr/src/main/java/org/apache/hadoop/mapreduce/RssMRConfig.java [hunk 9]
```diff
@@ -52,4 +53,15 @@ public class RssMRConfig {
   public static final String RSS_CLIENT_MAX_SEGMENT_SIZE = "mapreduce.rss.client.max.buffer.size";
   public static final long RSS_CLIENT_DEFAULT_MAX_SEGMENT_SIZE = 3 * 1024;
   public static final String RSS_STORAGE_TYPE = "mapreduce.rss.storage.type";
+
+  public static final String RSS_PARTITION_NUM_PER_RANGE = "mapreduce.rss.partitionNum.per.range";
+  public static final int RSS_PARTITION_NUM_PER_RANGE_DEFAULT_VALUE = 1;
+  public static final String RSS_BASE_PATH = "mapreduce.rss.base.path";
```
Comment (jerqi) [inline client-mr/src/main/java/org/apache/hadoop/mapreduce/RssMRConfig.java:59]: please concerned about the pr about mutli hdfs.

File: client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssBypassWriter.java (hunk 4 above)
Comment (jerqi) [reply client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssBypassWriter.java:32]: I need the comment about explaining why we need this class, and we don't use this class usually.

File: client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssBypassWriter.java [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+        + e.getMessage());
+    }
+    if (disk == null) {
+      throw new RssException("OnDiskMapOutput should not contain null disk stream");
+    }
+
+    // Copy data to local-disk
+    try {
+      disk.write(buffer.array(), 0, buffer.capacity());
+      disk.close();
```
Comment (jerqi) [inline client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssBypassWriter.java:58]: finally {
    disk.close();
}

File: client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssEventFetcher.java [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+import org.apache.hadoop.mapred.TaskCompletionEvent;
+import org.apache.hadoop.mapred.TaskUmbilicalProtocol;
+import org.apache.hadoop.mapreduce.MRRssUtils;
+import org.apache.hadoop.mapreduce.TaskAttemptID;
+import org.roaringbitmap.longlong.Roaring64NavigableMap;
+
+import com.tencent.rss.common.exception.RssException;
+
+public class RssEventFetcher<K,V> {
+  private static final long SLEEP_TIME = 1000;
```
Comment (jerqi) [inline client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssEventFetcher.java:38]: parameter from conf.

File: client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssEventFetcher.java [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+  public Roaring64NavigableMap fetchAllRssTaskIds() {
+    try {
+      acceptMapCompletionEvents();
+    } catch (Exception e) {
+      exceptionReporter.reportException(
+        new RssException("Reduce: " + reduce
+        + " fails to accept completion events due to: "
+        + e.getMessage())
+      );
+      return null;
```
Comment (jerqi) [inline client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssEventFetcher.java:78]: empty collection instead.

File: client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssEventFetcher.java [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+      if (!obsoleteMaps.contains(mapTask)) {
+        long taskId = MRRssUtils.convertTaskAttemptIdToLong(mapTask);
+        taskIds.addLong(taskId);
+      }
+    }
+    if (taskIds.getLongCardinality() + tipFailedCount != totalMapsCount) {
+      exceptionReporter.reportException(
+        new IllegalStateException("TaskAttemptIDs are inconsistent with total map tasks")
+      );
+      return null;
```
Comment (jerqi) [inline client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssEventFetcher.java:92]: ditto

File: client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssFetcher.java [hunk 14]
```diff
@@ -0,0 +1,237 @@
...
+
+  private final Reporter reporter;
+
+  private enum ShuffleErrors {
+    IO_ERROR, WRONG_LENGTH, BAD_ID, WRONG_MAP,
+    CONNECTION, WRONG_REDUCE
+  }
+
+  private static final String SHUFFLE_ERR_GRP_NAME = "Shuffle Errors";
+  private static final float BYTES_PER_MILLIS_TO_MBS = 1000f / 1024 / 1024;
```
Comment (jerqi) [inline client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssFetcher.java:54]: ByteUnit instead.

File: client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssShuffle.java [hunk 15]
```diff
@@ -0,0 +1,202 @@
...
+    this.umbilical = context.getUmbilical();
+    this.reporter = context.getReporter();
+    this.metrics = new ShuffleClientMetrics(reduceId, jobConf);
+    this.copyPhase = context.getCopyPhase();
+    this.taskStatus = context.getStatus();
+    this.reduceTask = context.getReduceTask();
+    merger = createMergeManager(context);
+
+    // rss init
+    String containerIdStr =
```
Comment (jerqi) [inline client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssShuffle.java:94]: Refactor a method in Util Class.

File: common/pom.xml (hunk 7 above)
Comment (frankliee) [reply common/pom.xml:74]: ok

File: client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssFetcher.java [hunk 16]
```diff
@@ -0,0 +1,237 @@
...
+        return;
+      }
+
+      // Write and commit uncompressed data to MapOutput.
+      // In the majority of cases, merger allocates memory to accept data,
+      // but when data size exceeds the threshold, merger can also allocate disk.
+      // So, we should consider the two situations, respectively.
+      try {
+        if (mapOutput instanceof InMemoryMapOutput) {
+          InMemoryMapOutput inMemoryMapOutput = (InMemoryMapOutput) mapOutput;
```
Comment (jerqi) [inline client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssFetcher.java:177]: put the judge logic into RssBypassWriter

File: client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssBypassWriter.java (hunk 10 above)
Comment (frankliee) [reply client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssBypassWriter.java:58]: Using finally is unnecessary, because disk.close can also throw IO Exception.
If there is an Exception on disk.write() or disk.close(), we just close this stream via MR's IOUtils.cleanup.

This code is similar to MR's OnDiskMapOutput.doShuffle

File: client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssFetcher.java (hunk 16 above)
Comment (frankliee) [reply client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssFetcher.java:177]: OK

File: client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssShuffle.java (hunk 15 above)
Comment (frankliee) [reply client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssShuffle.java:94]: ok

File: client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssFetcher.java (hunk 14 above)
Comment (frankliee) [reply client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssFetcher.java:54]: ok

File: client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssEventFetcher.java (hunk 12 above)
Comment (frankliee) [reply client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssEventFetcher.java:78]: ok

File: client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssEventFetcher.java (hunk 13 above)
Comment (frankliee) [reply client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssEventFetcher.java:92]: ok

File: client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssEventFetcher.java (hunk 11 above)
Comment (frankliee) [reply client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssEventFetcher.java:38]: This is useless, I will just remove it.

File: client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssFetcher.java [hunk 17]
```diff
@@ -0,0 +1,237 @@
...
+      long decompressDuration = System.currentTimeMillis() - startDecompress;
+      decompressTime += decompressDuration;
+
+      // Allocate a MapOutput (- either in-memory or on-disk) to put uncompressed block
+      final long startSerialization = System.currentTimeMillis();
+      TaskAttemptID mapId = MRRssUtils.createMRTaskAttemptId(
+        reduceId.getJobID(), TaskType.MAP, compressedBlock.getBlockId());
+      MapOutput<K, V> mapOutput = null;
+      try {
+        mapOutput = merger.reserve(mapId, compressedBlock.getUncompressLength(), 0);
```
Comment (jerqi) [inline client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssFetcher.java:158]: What's the beviour when the merger process the blocks which have the same mapIds.

Comment (frankliee) [reply client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssFetcher.java:158]: There will none same mapIDs, because they are generated from RSS blockIDs.

File: client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssFetcher.java [hunk 18]
```diff
@@ -0,0 +1,237 @@
...
+      ByteBuffer uncompressedData = RssShuffleUtils.decompressData(
+        compressedData, compressedBlock.getUncompressLength());
+      unCompressionLength += compressedBlock.getUncompressLength();
+      long decompressDuration = System.currentTimeMillis() - startDecompress;
+      decompressTime += decompressDuration;
+
+      // Allocate a MapOutput (- either in-memory or on-disk) to put uncompressed block
+      final long startSerialization = System.currentTimeMillis();
+      TaskAttemptID mapId = MRRssUtils.createMRTaskAttemptId(
+        reduceId.getJobID(), TaskType.MAP, compressedBlock.getBlockId());
```
Comment (jerqi) [inline client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssFetcher.java:155]: wrong method use.

File: client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssBypassWriter.java (hunk 10 above)
Comment (jerqi) [reply client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssBypassWriter.java:58]: you should 
try (disk) {
    
}

Comment (jerqi) [reply client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssBypassWriter.java:58]: It seems ok, but .... it's not normal. It make people think that you forget closing disk.

File: client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssFetcher.java (hunk 17 above)
Comment (jerqi) [reply client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssFetcher.java:158]: you should some comments, it's tricky. You should avoid this, if you have to do this. You should explain this with comments.

File: client-mr/src/main/java/org/apache/hadoop/mapreduce/MRRssUtils.java (hunk 8 above)
Comment (frankliee) [reply client-mr/src/main/java/org/apache/hadoop/mapreduce/MRRssUtils.java:2]: fixed

File: common/src/main/java/com/tencent/rss/common/RssShuffleUtils.java [hunk 19]
```diff
@@ -0,0 +1,51 @@
...
+import net.jpountz.lz4.LZ4FastDecompressor;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class RssShuffleUtils {
+
+  private static final Logger LOG = LoggerFactory.getLogger(RssShuffleUtils.class);
+
+  public static byte[] compressData(byte[] data) {
+    LZ4Compressor compressor = LZ4Factory.fastestInstance().fastCompressor();
```
Comment (jerqi) [inline common/src/main/java/com/tencent/rss/common/RssShuffleUtils.java:34]: MapOutputBuffer also replace with this method.

File: client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssBypassWriter.java [hunk 20]
```diff
@@ -0,0 +1,87 @@
...
+
+import com.tencent.rss.common.exception.RssException;
+
+// In MR shuffle, MapOutput encapsulates the logic to fetch map task's output data via http.
+// So, in RSS, we should bypass this logic, and directly write data to MapOutput.
+public class RssBypassWriter {
+  private static final Log LOG = LogFactory.getLog(RssBypassWriter.class);
+
+  public static void write(MapOutput mapOutput, ByteBuffer buffer,
+                           TaskAttemptID reduceId, long rssBlockId) {
```
Comment (jerqi) [inline client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssBypassWriter.java:39]: I think we should remove blockId, and Compressed data also remove blockId.

File: client-mr/src/main/java/org/apache/hadoop/mapreduce/RssMRConfig.java (hunk 9 above)
Comment (frankliee) [reply client-mr/src/main/java/org/apache/hadoop/mapreduce/RssMRConfig.java:59]: The mutli-hdfs pr does not change the conf key of "rss.base.path".

Comment (jerqi) [reply client-mr/src/main/java/org/apache/hadoop/mapreduce/RssMRConfig.java:59]: > The mutli-hdfs pr does not change the conf key of "rss.base.path".

you should guarantee multi hdfs's pr can work with your implement.

File: client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssBypassWriter.java (hunk 20 above)
Comment (frankliee) [reply client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssBypassWriter.java:39]: you mean "write(MapOutput mapOutput, ByteBuffer buffer)"?

File: common/src/main/java/com/tencent/rss/common/RssShuffleUtils.java (hunk 19 above)
Comment (frankliee) [reply common/src/main/java/com/tencent/rss/common/RssShuffleUtils.java:34]: Ok

File: client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssFetcher.java (hunk 18 above)
Comment (frankliee) [reply client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssFetcher.java]: Update to use a faked TaskAttemptId instead.

File: client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssFetcher.java (hunk 17 above)
Comment (frankliee) [reply client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssFetcher.java:163]: The comment is added.

File: client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssBypassWriter.java (hunk 20 above)
Comment (jerqi) [reply client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssBypassWriter.java:39]: Yes

Comment (frankliee) [reply client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssBypassWriter.java]: Done

File: client-mr/src/main/java/org/apache/hadoop/mapreduce/RssMRConfig.java (hunk 9 above)
Comment (frankliee) [reply client-mr/src/main/java/org/apache/hadoop/mapreduce/RssMRConfig.java:59]: In multi hdfs's pr, the interface of ShuffleClientFactory is unchanged.

File: client-spark/common/src/main/java/org/apache/spark/shuffle/RssSparkShuffleUtils.java [hunk 21]
```diff
@@ -45,9 +41,9 @@
 import com.tencent.rss.client.factory.CoordinatorClientFactory;
 import com.tencent.rss.storage.util.StorageType;
 
-public class RssShuffleUtils {
+public class RssSparkShuffleUtils {
```
Comment (colinmjj) [inline client-spark/common/src/main/java/org/apache/spark/shuffle/RssSparkShuffleUtils.java:44]: The name refactor should be included in another PR.
Too many code in spark module are updated, but this PR is target to MR's reduce part.

File: client/src/main/java/com/tencent/rss/client/response/CompressedShuffleBlock.java [hunk 22]
```diff
@@ -37,4 +39,8 @@ public int getUncompressLength() {
   public ByteBuffer getByteBuffer() {
     return byteBuffer;
   }
+
+  public long getBlockId() {
```
Comment (jerqi) [inline client/src/main/java/com/tencent/rss/client/response/CompressedShuffleBlock.java:43]: remove.

Comment (frankliee) [reply client/src/main/java/com/tencent/rss/client/response/CompressedShuffleBlock.java:43]: Without it, is it impossible to know which block is fetched from `readShuffleBlockData()`.
And this is helpful for debug.

File: client-spark/common/src/main/java/org/apache/spark/shuffle/RssSparkShuffleUtils.java (hunk 21 above)
Comment (frankliee) [reply client-spark/common/src/main/java/org/apache/spark/shuffle/RssSparkShuffleUtils.java:44]: This is because that the original RssShuffleUtils contains some spark-related dependency,  which should not be used by MR.  So this PR has to move it from spark, and put it into common.
Otherwise, MR needs to the dependency of Spark, it looks more strange.

File: client/src/main/java/com/tencent/rss/client/response/CompressedShuffleBlock.java (hunk 22 above)
Comment (jerqi) [reply client/src/main/java/com/tencent/rss/client/response/CompressedShuffleBlock.java:43]: I don't think that we should add this field for debug.

Comment (frankliee) [reply client/src/main/java/com/tencent/rss/client/response/CompressedShuffleBlock.java]: Removed.

File: client-mr/src/main/java/org/apache/hadoop/mapreduce/RssMRConfig.java (hunk 9 above)
Comment (jerqi) [reply client-mr/src/main/java/org/apache/hadoop/mapreduce/RssMRConfig.java]: I see that.

Comment (colinmjj) [reply client-mr/src/main/java/org/apache/hadoop/mapreduce/RssMRConfig.java]: for client config, there should be a PR to optimize config name. eg,
in coordinator:  rss.client.aaa.bbb
in spark: spark.rss.aaa.bbb
in mr: mapreduce.rss.aa.bbb

**Comment by jerqi** [comment]
LGTM, If GA pass, I will approve it.

**Review by jerqi — DISMISSED** [review]

**Review by jerqi — APPROVED** [review]
